### PR TITLE
Add npartitions to head

### LINKED ISSUE #506: Add npartitions to head and tail
Add the arg

### PR DESCRIPTION
**Pull request by milesgranger**
Will close #506 

Dask doesn't have `npartitions` for `.tail` mentioned in the original issue, so I did not add it here either.

### CODE REVIEW

File: dask_expr/_collection.py [hunk 1]
```diff
@@ -346,8 +346,8 @@ def index(self, value):
     def reset_index(self, drop=False):
         return new_collection(expr.ResetIndex(self, drop))
 
-    def head(self, n=5, compute=True):
-        out = new_collection(expr.Head(self, n=n))
+    def head(self, n=5, compute=True, npartitions=1):
```
Comment (phofl) [inline dask_expr/_collection.py:349]: Can you switch npartitions and compute to avoid changing the argument order?

File: dask_expr/_expr.py [hunk 2]
```diff
@@ -1595,9 +1595,20 @@ def _simplify_up(self, parent, dependents):
     def _lower(self):
         if not isinstance(self, BlockwiseHead):
             # Lower to Blockwise
+            if hasattr(self.frame, "_partitions") and self.npartitions > len(
```
Comment (phofl) [inline dask_expr/_expr.py:1598]: This goes back to the class PartitionsFiltered, so you can do an isinstance check

File: dask_expr/_expr.py [hunk 3]
```diff
@@ -1595,9 +1595,20 @@ def _simplify_up(self, parent, dependents):
...
+                self.frame._partitions
+            ):
+                raise ValueError(
+                    f"only {len(self.frame._partitions)} partitions, head received {self.npartitions}"
+                )
+
+            if hasattr(self.frame, "_partitions"):
+                partitions = self.frame._partitions[: self.npartitions]
```
Comment (phofl) [inline dask_expr/_expr.py:1606]:
```suggestion
                partitions = self.frame._partitions[: self.operand("npartitions")]
```

File: dask_expr/_expr.py [hunk 4]
```diff
@@ -1595,9 +1595,20 @@ def _simplify_up(self, parent, dependents):
...
+            else:
+                partitions = [0]
```
[the remark this answers is not in the document]
Comment (phofl) [reply dask_expr/_expr.py:1608]: You would use list(range(self.frame.npartitions)) here

[the remark this answers is not in the document]
Comment (phofl) [reply dask_expr/_expr.py:1608]: and then slice this list

**Comment by phofl** [comment]
Can you open an issue here for tail also adding npartitions? This isn't urgent, so no need to label with a Milestone, but we should keep track of it

File: dask_expr/_collection.py (hunk 1 above)
Comment (milesgranger) [reply dask_expr/_collection.py]: Oh ya, good catch. https://github.com/dask-contrib/dask-expr/pull/636/commits/536fc36e22578ee79f37004452fbb7a7ffdd63f0

File: dask_expr/_expr.py (hunk 2 above)
Comment (milesgranger) [reply dask_expr/_expr.py]: https://github.com/dask-contrib/dask-expr/pull/636/commits/63d15469bf460e969021f55f001c10e3b657f9d5, thanks!

**Comment by milesgranger** [comment]
> Can you open an issue here for tail also adding npartitions? This isn't urgent, so no need to label with a Milestone, but we should keep track of it.

https://github.com/dask-contrib/dask-expr/issues/639

File: dask_expr/_expr.py [hunk 5]
```diff
@@ -1595,9 +1595,19 @@ def _simplify_up(self, parent, dependents):
     def _lower(self):
         if not isinstance(self, BlockwiseHead):
             # Lower to Blockwise
+            if self.npartitions > self.frame.npartitions:
```
Comment (phofl) [inline dask_expr/_expr.py:1598]: operand(npartitions) as well here please

File: dask_expr/_expr.py [hunk 6]
```diff
@@ -1595,9 +1595,19 @@ def _simplify_up(self, parent, dependents):
...
+                raise ValueError(
+                    f"only {self.frame.npartitions} partitions, head received {self.npartitions}"
+                )
+
+            if isinstance(self, PartitionsFiltered):
+                partitions = self.frame._partitions[: self.operand("npartitions")]
+            else:
+                partitions = list(range(self.frame.npartitions)[: self.npartitions])
```
Comment (phofl) [inline dask_expr/_expr.py:1606]: Sorry my recommendation was pretty dumb, you can just do 

```
list(range(self.operand("npartitions"))
```

Comment (milesgranger) [reply dask_expr/_expr.py]: haha, I totally just did that and didn't even register it. :laughing:

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx @milesgranger

**Pull request merged**